render/pixman/renderer.c

```diff
@@ -0,0 +1,410 @@
+
+	wlr_log(WLR_INFO, "tex is w %zu h %zu", (size_t)wlr_texture->width,
+			(size_t)wlr_texture->height);
+
+	wlr_log(WLR_INFO, "stride is %zu", (size_t)stride);
+
+	wlr_log(WLR_INFO, "src_x %zu src_y %zu", (size_t)src_x, (size_t)src_y);
+
+	uint32_t *start = (uint32_t*)data + (src_y * stride) + (src_x * 4);
+	assert(start);
+
+	pixman_image_t *pixels = pixman_image_create_bits_no_clear(texture->format,
+			width, height, start, stride);
+
+	pixman_image_composite32(PIXMAN_OP_OVER, texture->image, NULL, pixels,
+			dst_x, dst_y, 0, 0, 0, 0, width, height);
+
+	pixman_image_unref(pixels);
+
+	return true;
+}
+
+static bool texture_to_dmabuf(struct wlr_texture *wlr_texture,
+		struct wlr_dmabuf_attributes *attribs) {
+	wlr_log(WLR_ERROR, "todo texture_to_dmabuf");
+	return false;
+}
+
+static void texture_destroy(struct wlr_texture *wlr_texture) {
+	if (wlr_texture == NULL) {
+		return;
+	}
+	struct wlr_pixman_texture *texture = get_texture(wlr_texture);
+
+	pixman_image_unref(texture->image);
+	free(texture);
+}
+
+static const struct wlr_texture_impl texture_impl = {
+	.is_opaque = texture_is_opaque,
+	.write_pixels = texture_write_pixels,
+	.to_dmabuf = texture_to_dmabuf,
+	.destroy = texture_destroy,
+};
+
+static uint32_t get_pixman_format_from_wl(enum wl_shm_format wl_fmt) {
+	uint32_t fmt = 0;
+	switch (wl_fmt) {
+	case WL_SHM_FORMAT_ARGB8888:
+#if WLR_BIG_ENDIAN
+		fmt = PIXMAN_b8g8r8a8;
+#else
+		fmt = PIXMAN_a8r8g8b8;
+#endif
+		break;
+	case WL_SHM_FORMAT_XRGB8888:
+#if WLR_BIG_ENDIAN
+		fmt = PIXMAN_b8g8r8x8;
+#else
+		fmt = PIXMAN_x8r8g8b8;
+#endif
+		break;
+	case WL_SHM_FORMAT_ABGR8888:
+#if WLR_BIG_ENDIAN
+		fmt = PIXMAN_r8g8b8a8;
+#else
+		fmt = PIXMAN_a8b8g8r8;
+#endif
+		break;
+	default:
+		wlr_log(WLR_ERROR, "Unsupported wl_shm_format 0x%"PRIX32, wl_fmt);
+		break;
+	}
+
+	return fmt;
+}
+
+struct wlr_pixman_texture *pixman_create_texture(
+		struct wlr_texture *wlr_texture, struct wlr_pixman_renderer *renderer);
+
+static void handle_destroy_buffer(struct wl_listener *listener, void *data) {
+	struct wlr_pixman_buffer *buffer =
+		wl_container_of(listener, buffer, buffer_destroy);
+	wl_list_remove(&buffer->link);
+	wl_list_remove(&buffer->buffer_destroy.link);
+
+	pixman_image_unref(buffer->image);
+
+	free(buffer);
+}
+
+static struct wlr_pixman_buffer *create_buffer(
+		struct wlr_pixman_renderer *renderer, struct wlr_buffer *wlr_buffer) {
+	struct wlr_pixman_buffer *buffer = calloc(1, sizeof(*buffer));
+	if (buffer == NULL) {
+		wlr_log_errno(WLR_ERROR, "Allocation failed");
+		return NULL;
+	}
+	buffer->buffer = wlr_buffer;
+	buffer->renderer = renderer;
+
+	uint32_t drm_format;
+	struct wlr_dmabuf_attributes dmabuf = {0};
+	struct wlr_shm_attributes shm = {0};
+	if (wlr_buffer_get_dmabuf(wlr_buffer, &dmabuf)) {
+		drm_format = dmabuf.format;
+	} else if (wlr_buffer_get_shm(wlr_buffer, &shm)) {
+		drm_format = shm.format;
+	} else {
+		goto error_buffer;
+	}
+
+	uint32_t format;
```

Comment:
We could sanity-check the format with `pixman_format_supported_destination()` and print a good error message if it returns false.